csi/controller.go

```diff
@@ -315,11 +316,154 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
 	}
 }
 
+// CreateVolume is a CSI API which creates a volume on OSD
+// This function supports snapshots if the parent volume id is supplied
+// in the parameters.
+func (s *OsdCsiServer) CreateVolume(
+	ctx context.Context,
+	req *csi.CreateVolumeRequest,
+) (*csi.CreateVolumeResponse, error) {
+
+	// Log request
+	dlog.Debugf("CreateVolume req[%#v]", *req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name must be provided")
+	}
+	if req.GetVolumeCapabilities() == nil || len(req.GetVolumeCapabilities()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities must be provided")
+	}
+	if req.GetCapacityRange() == nil {
+		return nil, status.Error(codes.InvalidArgument, "CapacityRange must be specified")
+	}
+	if req.GetCapacityRange().GetRequiredBytes() == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Capacity range required bytes cannot be zero")
+	}
+
+	// Create response
+	volume := &csi.VolumeInfo{}
+	resp := &csi.CreateVolumeResponse{
+		Reply: &csi.CreateVolumeResponse_Result_{
+			Result: &csi.CreateVolumeResponse_Result{
+				VolumeInfo: volume,
+			},
+		},
+	}
+
+	// Check if the volume has already been created or is in process of creation
+	v, err := util.VolumeFromName(s.driver, req.GetName())
+	if err == nil {
+		osdToCsiVolumeInfo(volume, v)
+		return resp, nil
+	}
+
+	// Get parameters
+	spec, locator, source, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		e := fmt.Sprintf("Unable to get parameters: %s\n", err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.InvalidArgument, e)
+	}
+
+	// Check if the caller is asking to create a snapshot or for a new volume
+	var id string
+	if source != nil && len(source.GetParent()) != 0 {
```

Comment:
This should be left to the drivers to handle ?

Reply:
That's how it works for the docker manager in osd.